subprojects/language-native/src/main/java/org/gradle/language/cpp/plugins/CppApplicationPlugin.java

```diff
@@ -136,7 +137,7 @@ public String call() throws Exception {
 
                             // Use the debug variant as the development binary
                             // Prefer the host architecture, if present, else use the first architecture specified
-                            if (buildType == BuildType.DEBUG && (targetMachine.getArchitecture().equals(targetMachineFactory.host().getArchitecture()) || !application.getDevelopmentBinary().isPresent())) {
+                            if (buildType == BuildType.DEBUG && (targetMachine.getArchitecture().equals(((DefaultTargetMachineFactory)targetMachineFactory).host().getArchitecture()) || !application.getDevelopmentBinary().isPresent())) {
```

Comment:
Should avoid the cast, and instead inject the factory as the type that the plugin needs (`DefaultTargetMachineFactory`).